src/lib/update-lock.ts

```diff
@@ -96,69 +87,93 @@ function dispose(
-								return Bluebird.mapSeries(
-									lockFilesOnHost(appId, serviceName),
-									(tmpLockName) => {
-										return (
-											Bluebird.try(() => {
-												if (force || lockOverride) {
-													return lockfile.unlock(tmpLockName);
-												}
-											})
-												.then(() => {
-													return lockfile.lock(tmpLockName, LOCKFILE_UID);
-												})
-												// If lockfile exists, throw a user-friendly error.
-												// Otherwise throw the error as-is.
-												// This will interrupt the call to Bluebird.using, so
-												// dispose needs to be called even though it's referenced
-												// by .disposer later.
-												.catch((error) => {
-													return dispose(release, appId).throw(
-														lockfile.LockfileExistsError.is(error)
-															? new UpdatesLockedError(
-																	`Lockfile exists for ${JSON.stringify({
-																		serviceName,
-																		appId,
-																	})}`,
-															  )
-															: (error as Error),
-													);
-												})
-										);
-									},
-								);
-							});
-					})
-					.disposer((release: () => void) => dispose(release, appId));
-			})
-			.catch((err) => {
-				throw new InternalInconsistencyError(
-					`Error getting lockOverride config value: ${err?.message ?? err}`,
-				);
-			});
-	};
+	// Sort appIds so they are always locked in the same sequence
+	const sortedIds = appIdsToLock.sort();
 
-	const disposer = takeTheLock();
+	let lockOverride: boolean;
+	try {
+		lockOverride = await config.get('lockOverride');
+	} catch (err) {
+		throw new InternalInconsistencyError(
+			`Error getting lockOverride config value: ${err?.message ?? err}`,
+		);
+	}
 
-	return Bluebird.using(disposer, fn as () => PromiseLike<T>);
+	const releases = new Map<number, () => void>();
+	try {
+		for (const id of sortedIds) {
+			const lockDir = getPathOnHost(lockPath(id));
+			// Acquire write lock for appId
+			releases.set(id, await writeLock(id));
+			// Get list of service folders in lock directory
+			let serviceFolders: string[] = [];
+			try {
+				serviceFolders = await fs.readdir(lockDir);
+			} catch (err) {
+				if (ENOENT(err)) {
+					// Default to empty list of folders
+					serviceFolders = [];
+				} else {
+					throw err;
+				}
+			}
+			// Attempt to create a lock for each service
+			for (const service of serviceFolders) {
+				await lockService(id, service, { force, lockOverride });
+			}
+		}
+		// Resolve the function passed
+		return Promise.resolve(fn());
+	} finally {
+		// Cleanup locks
+		for (const [id, release] of releases.entries()) {
+			// Release lock on application
+			release();
+			// Dispose of locks in each service within the application
+			await dispose(id);
+		}
+	}
+}
+
+type LockOptions = {
+	force?: boolean;
+	lockOverride?: boolean;
+};
+
+async function lockService(
+	appId: number,
+	service: string,
+	opts: LockOptions = {
+		force: false,
```

Comment:
Do you need the two options, `force` and `lockOverride` here? Since this function is only called internally within the module, you could just define a `force` option and call it as `lockService(appId, service, force || lockOverride)`

Reply:
Maybe adding a catch all within `dispose()` will be enough

Reply:
We want to remove lockOverride or force since they are basically duplicates but I didn't want to do that in this PR. I just copied exactly how the previous code worked which wanted to evaluated both conditions https://github.com/balena-os/balena-supervisor/blob/33ad877c67a639024a35cd4aa8452d280f757a45/src/lib/update-lock.ts#L123

Reply:
I realized after doing `lockService(appId, service, force || lockOverride)` provides the same logic while improving the lockService interface so I did it!